Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+
+    Note left of Lumberjack: Should use same NATS subject
+    
+    Note right of Lumberjack: Uses pino-elasticsearch
+```
+
+
+At the heart of Tazama's logging stack, there is [pino], a fast json logger which is highly configurable. Tazama also uses the ELK stack for centralised logging of the logs generated by different microservices. There are 3 main components involved in the stack:
+
+- logger
+- [lumberjack]
+- [event-sidecar]
+- [nats]
+
+
+#### Logger
+Tazama uses [pino] as the core logger. Which, at its core is a `json` logger. Pino utilises **transports** [^transport], which describe components that can be used to transmit and transform log output. Pino [recommends](https://github.com/pinojs/pino/blob/main/docs/transports.md#transports) that:
+
+> ... any log transformation or transmission is performed either in a separate thread or a separate process.
+
+Which is where the [event-sidecar] comes in:
+
+#### Event-Sidecar
+As per the [pino] recommendation, the [event-sidecar] is a microservice that runs alongside a processor, capturing log events and transmitting them to a set destination [^lumberjack]. 
+
+As the [event-sidecar] is running in a separate process, some inter-process-communication is required in order to transmit the logs from the main processor, to the [event-sidecar]. This is implemented using [gRPC], which involves two pieces. A [client](#client) and a [server](#server).
+
+##### Server
+The [event-sidecar] itself, is a [gRPC] server that listens for requests sent by compatible clients following a specific format. The message sent is defined in a protobuf [format][wire]. An extract:
+
+```proto
+enum LogLevel {
+  trace = 0;
+  // omitted
+  fatal = 5;
+}
+message LogMessage {
+  string message = 1; // A message for the log
+  LogLevel level = 2; // Log level
+  // omitted
+}
+
+service Lumberjack {
+  // Take a LogMessage object and an empty message
+  rpc SendLog (LogMessage) returns (google.protobuf.Empty);
+}
+```
+
+##### Client
+Tazama abstracts away the need to write your own [gRPC] client. Simply initialising a logger and providing it with an address will create a [gRPC] client behind the scenes which will send requests to the address that was provided.
+
+An example:
+
+```js
+// No sidecar host is provided in the LoggerService constructor
+const logger = new LoggerService();
+```
+
+An alternative implementation, **opting into** logging with the sidecar (as well as [lumberjack]) may be:
+```js
+const logger = new LoggerService("192.0.0.1:1234");
+```
+> [!NOTE]  
+> If a sidecar host is provided in the `LoggerService` constructor, you have to ensure your sidecar is live on the address you provided
+
+As [gRPC] is in use, it allows Tazama to be flexible enough to allow logs coming from any client (in any programming language) as long as it adheres to the type that the server expects (defined in the protobuf format above).
+
+After receiving the logs, the [event-sidecar] sends them to [lumberjack] via [NATS]
```

Comment:
> it would also make sense if a separate logging-specific instance of the NATS server is deployed so that you do not have contention between the logging activity and the "operational" activity and a possible negative impact on base system performance just because logging is enabled

I think we should keep this generic. All we need is a NATS server. I don't think where that comes from or where it's being used should be in the scope of these docs as it ultimately does not matter in getting the logging framework setup.

> It would also be good to understand how the logging is set up to use NATS (subjects, environment variables, etc).

I do have a section on this. I will try to restructure the docs so they are easy to follow.